openassessment/xblock/utils/allow_resubmission.py

```diff
@@ -0,0 +1,97 @@
+
+from datetime import datetime, timedelta
+import pytz
+
+
+def allow_resubmission(config_data, workflow_data, submission_data: dict) -> bool:
+    """
+    Determines if a learner can reset their submission and try again. A learner
+    can resubmit if the following conditions are met:
+    - Whether the assignment allows learner resubmissions
+    - Whether the submission date has not been exceeded
+    - Whether the learner's response has not been graded by peers or staff
+
+    Args:
+        config_data (ORAConfigAPI): Object with all the configuration data of the ORA assignment.
+        workflow_data (WorkflowAPI): Object containing necessary data for time and workflow status checks.
+        submission_data (dict): Dictionary containing the submission data of the learner.
+
+    Returns:
+        bool: True if the learner's submission is eligible for retry, False otherwise.
+    """
+    return (
+        allow_learner_resubmissions(config_data) and not
+        submission_date_exceeded(config_data, submission_data) and not
+        has_been_graded(workflow_data)
+    )
+
+
+def allow_learner_resubmissions(config_data) -> bool:
+    """
+    Checks if the ORA assignment allows learner resubmissions.
+
+    Args:
+        config_data (ORAConfigAPI): Object with all the configuration data of the ORA assignment.
+
+    Returns:
+        bool: True if the assignment allows learner resubmissions, False otherwise
+    """
+    return config_data.allow_learner_resubmissions
+
+
+def submission_date_exceeded(config_data, submission_data: dict) -> bool:
+    """
+    Checks if the submission due date has been exceeded and if the learner is
+    within the grace period. The grace period is the time after the submission
+    date that the learner can still resubmit their response.
+
+    If the grace period is set to 0 hours and 0 minutes, means that the learner
+    can resubmit their response at any time before the submission due date.
+
+    Args:
+        config_data (ORAConfigAPI): Object with all the configuration data of the ORA assignment.
+        submission_data (dict): Dictionary containing the submission data of the learner.
+
+    Returns:
+        bool: True if the submission date has been exceeded, False otherwise.
+    """
+    is_closed, reason, _, _ = config_data.is_closed(step="submission")
+    if is_closed and reason == "due":
+        return True
+
+    hours = config_data.resubmissions_grace_period_hours
+    minutes = config_data.resubmissions_grace_period_minutes
+    if not hours and not minutes:
+        return False
+
+    current_datetime = datetime.now(pytz.UTC)
+    grace_period = timedelta(hours=int(hours), minutes=int(minutes))
+    deadline_datetime = submission_data["created_at"] + grace_period
+    return current_datetime >= deadline_datetime
+
+
+def has_been_graded(workflow_data) -> bool:
+    """
+    Checks if the learner's submission hasn't been graded by peers or staff.
```

Comment:
Question: What should happen when an instructor starts the grading process but hasn't finished, and a student tries resubmitting?

Reply:
Same question about peers.

Reply:
I believe this behavior changed, now peer assessments are not accepted.

Reply:
Yes, with the new changes, these messages no longer appear because:
1. If a peer step is included, no student can reset.
2. If an instructor has already started a student's grade, that student cannot reset.